File: search in 2d matrix.py
#Time complexity: O(M+N) M and N denote number of rows and columns
#Space complexity: O(1)
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        i=0
        j=len(matrix[0])-1
        while i<len(matrix) and j>=0:
            #Checking if we found our target
            if matrix[i][j]==target:
                return True
            elif matrix[i][j]>target:
                #It means the values below are greater than the target and hence cn be ignored
                j=j-1
            elif matrix[i][j]<target:
                #It means the values to the left are smaller and can be ignored.
                i=i+1
        return False

File: Searchin2DMatrix.cpp
//Time complexity: O(log(m*n)) M=number of rows 
// N=no of columns -> Hypothetically converting 
//2D matrix to linear array
//Space : O(1)

class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        int m=matrix.size();
        int n= matrix[0].size();
        int left=0, right=m*n-1;

       
        while(left<=right)
        {
                int mid=left+(right-left)/2;
                int row=mid/n, col=mid%n;
                int num=matrix[row][col];
            if(target==num)
            {
                return true;

            }
             if(target<num)
            {
                right=mid-1;
            }
            else
            {
                left=mid+1;

            }
        }
        return false;
    }
};